include/boost/hana/detail/operators/comparable.hpp

```diff
@@ -24,17 +24,29 @@ BOOST_HANA_NAMESPACE_BEGIN namespace detail {
         static constexpr bool value = false;
     };
 
+    template <typename Tag, typename T>
+    struct comparable_operators_sfinae_helper
+        : comparable_operators<Tag>
+    { };
+
+    template <typename Tag>
+    struct comparable_operators_sfinae_helper<Tag, Tag>;
+
+    template <typename T>
+    using comparable_operators_sfinae = comparable_operators_sfinae_helper<
+        hana::tag_of_t<T>, std::remove_const_t<std::remove_reference_t<T>>>;
+
     namespace operators {
         template <typename X, typename Y, typename = typename std::enable_if<
-            detail::comparable_operators<typename hana::tag_of<X>::type>::value ||
-            detail::comparable_operators<typename hana::tag_of<Y>::type>::value
+            detail::comparable_operators_sfinae<X>::value ||
```

Comment:
How about:

```
namespace detail {
  template <typename T>
  struct has_idempotent_tag
    : std::is_same<hana::tag_of_t<T>, std::remove_const_t<std::remove_reference_t<T>>>
  { };
}
```

And then

```
std::enable_if_t<(!detail::has_idempotent_tag<X>::value && detail::comparable_operators<X>::value) ||  ...>
```

If we want to be lazy, we can define (or perhaps we already have somewhere) a lazy `and_` type trait. WDYT?

Reply:
I'm not sure what you mean by "lazy". Do you mean short-circuiting the template instantiations?

Reply:
I used your metafunction, but the logic was a bit different. Neither of the operands could be a tag, or for some reason it compiled and always returned false because it was finding the default impl. (I think)